src/dex/maverick-v2/maverick-v2-pool.ts

```diff
@@ -0,0 +1,279 @@
+    this.addressesSubscribed = [address];
+
+    // Add handlers
+    this.handlers['PoolAddLiquidity'] = this.handleAddLiquidityEvent.bind(this);
+    this.handlers['PoolRemoveLiquidity'] =
+      this.handleRemoveLiquidityEvent.bind(this);
+    this.handlers['PoolSwap'] = this.handleSwapEvent.bind(this);
+
+    this.poolMath = new MaverickPoolMath(
+      feeAIn,
+      feeBIn,
+      lookback,
+      tickSpacing,
+      protocolFeeRatio,
+      activeTick,
+      BigInt(tokenA.decimals),
+      BigInt(tokenB.decimals),
+    );
+
+    this.poolContract = new this.dexHelper.web3Provider.eth.Contract(
+      MaverickV2PoolABI as AbiItem[],
+      this.address,
+    );
+
+    this.poolLensContract = new this.dexHelper.web3Provider.eth.Contract(
+      MaverickV2PoolLensABI as AbiItem[],
+      this.poolLensAddress,
+    );
+  }
+
+  /**
+   * The function is called every time any of the subscribed
+   * addresses release log. The function accepts the current
+   * state, updates the state according to the log, and returns
+   * the updated state.
+   * @param state - Current state of event subscriber
+   * @param log - Log released by one of the subscribed addresses
+   * @returns Updates state of the event subscriber after the log
+   */
+  protected processLog(
+    state: DeepReadonly<PoolState>,
+    log: Readonly<Log>,
+    blockHeader: Readonly<BlockHeader>,
+  ): DeepReadonly<PoolState> | null {
+    try {
+      const event = this.logDecoder(log);
+      if (event.name in this.handlers) {
+        return this.handlers[event.name](event, state, log, blockHeader);
+      }
+    } catch (e) {
+      catchParseLogError(e, this.logger);
+    }
+
+    return null;
+  }
+
+  /**
+   * The function generates state using on-chain calls. This
+   * function is called to regenerate state if the event based
+   * system fails to fetch events and the local state is no
+   * more correct.
+   * @param blockNumber - Blocknumber for which the state should
+   * should be generated
+   * @returns state of the event subscriber at blocknumber
+   */
+
+  async generateState(blockNumber: number): Promise<DeepReadonly<PoolState>> {
+    try {
+      const poolContractState = await this.poolContract.methods
+        .getState()
+        .call({}, blockNumber);
+
+      const poolState: PoolState = {
+        activeTick: BigInt(poolContractState.activeTick),
+        binCounter: BigInt(poolContractState.binCounter),
+        reserveA: BigInt(poolContractState.reserveA),
+        reserveB: BigInt(poolContractState.reserveB),
+        lastTwaD8: BigInt(poolContractState.lastTwaD8),
+        lastLogPriceD8: BigInt(poolContractState.lastLogPriceD8),
+        lastTimestamp: BigInt(poolContractState.lastTimestamp),
+        bins: {},
+        ticks: {},
+      };
+
+      for (let i = 0n; i < poolState.binCounter / 5000n + 1n; i++) {
+        const poolLensState = await this.poolLensContract.methods[
```

Comment:
@Wertusser Would be more efficient to use multi call here

Reply:
@aburkut fixed